File: server/src/services/ai/storyGenerator.js
import { ai, models } from './geminiClient.js';
import { buildWorldPrompt } from './prompts/worldBuilding.js';
import { withRetry } from './retry.js';
import logger from '../../utils/logger.js';

export const storyGenerator = {
  async generateWorld(setup) {
    const prompt = buildWorldPrompt(setup);

    logger.info('Generating world with Gemini Pro...');
    const response = await withRetry(() => ai.models.generateContent({
      model: models.storyPro,
      contents: [{ role: 'user', parts: [{ text: prompt }] }],
      config: {
        temperature: 0.9,
        maxOutputTokens: 65536,
        responseMimeType: 'application/json',
      },
    }), { label: 'storyGenerator.generateWorld' });

    // Concatenate all text parts from the response
    let text = '';
    for (const part of response.candidates[0].content.parts) {
      if (part.text) text += part.text;
    }

    let parsed;
    try {
      parsed = JSON.parse(text);
    } catch (e) {
      // Try extracting JSON from markdown code blocks
      const match = text.match(/```(?:json)?\s*([\s\S]*?)```/);
      if (match) {
        try {
          parsed = JSON.parse(match[1]);
        } catch (e2) {
          logger.error('Failed to parse extracted JSON:', e2.message);
        }
      }

      if (!parsed) {
        // Try to find first { and last } to extract JSON
        const firstBrace = text.indexOf('{');
        const lastBrace = text.lastIndexOf('}');
        if (firstBrace !== -1 && lastBrace > firstBrace) {
          try {
            parsed = JSON.parse(text.slice(firstBrace, lastBrace + 1));
          } catch (e3) {
            logger.error('All JSON parse attempts failed. Response length:', text.length);
            logger.error('Response start:', text.slice(0, 300));
            logger.error('Response end:', text.slice(-300));
            throw new Error('Failed to parse story generation response as JSON');
          }
        } else {
          logger.error('No JSON object found in response. Length:', text.length);
          throw new Error('Failed to parse story generation response as JSON');
        }
      }
    }

    // Validate required fields
    if (!parsed.worldLore || !parsed.characters || !parsed.initialScenes || !parsed.openingScript) {
      const missing = ['worldLore', 'characters', 'initialScenes', 'openingScript']
        .filter((k) => !parsed[k]);
      throw new Error(`Story generation response missing required fields: ${missing.join(', ')}`);
    }

    if (parsed.characters.length < 3) {
      logger.warn(`Only ${parsed.characters.length} characters generated, expected 3-5`);
    }

    logger.info(
      `World generated: ${parsed.worldLore.name}, ${parsed.characters.length} characters, ${parsed.initialScenes.length} scenes`
    );

    return parsed;
  },
};

File: server/app/services/ai/story_generator.py
import json
import re
from typing import Any

from google.genai import types

from app.logger import logger
from app.services.ai.gemini_client import get_client, models
from app.services.ai.prompts.world_building import build_continuation_prompt, build_world_prompt


def _parse_json_loose(text: str) -> Any:
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        pass

    # Try fenced markdown
    fence = re.search(r"```(?:json)?\s*([\s\S]*?)```", text)
    if fence:
        try:
            return json.loads(fence.group(1))
        except json.JSONDecodeError as e2:
            logger.error("Failed to parse extracted JSON: %s", e2)

    # Fall back to first { ... last } slice
    first = text.find("{")
    last = text.rfind("}")
    if first != -1 and last > first:
        try:
            return json.loads(text[first : last + 1])
        except json.JSONDecodeError:
            pass

    logger.error("All JSON parse attempts failed. Length=%d", len(text))
    logger.error("Response start: %s", text[:300])
    logger.error("Response end:   %s", text[-300:])
    raise ValueError("Failed to parse story generation response as JSON")


def generate_world(setup: dict, *, continuation: dict | None = None) -> dict:
    """Generate a complete world. If `continuation` is provided, treat this as
    the next chapter of an existing story — same world + cast, new spine +
    endings, picking up after the previous chapter's chosen ending.

    `continuation` keys: prior_session (row dict), prior_world (dict),
    prior_ending (dict), chapter_number (int).
    """
    if continuation:
        prompt = build_continuation_prompt(
            prior=continuation["prior_session"],
            prior_world=continuation["prior_world"],
            prior_ending=continuation["prior_ending"],
            chapter_number=continuation["chapter_number"],
        )
    else:
        prompt = build_world_prompt(setup)
    client = get_client()

    logger.info(
        "Generating %s with Gemini Pro...",
        "continuation chapter" if continuation else "world",
    )
    response = client.models.generate_content(
        model=models.story_pro,
        contents=[
            types.Content(role="user", parts=[types.Part.from_text(text=prompt)])
        ],
        config=types.GenerateContentConfig(
            temperature=0.9,
            max_output_tokens=65536,
            response_mime_type="application/json",
        ),
    )

    text = ""
    for part in response.candidates[0].content.parts:
        if getattr(part, "text", None):
            text += part.text

    parsed = _parse_json_loose(text)

    required = ["worldLore", "characters", "scenes", "storySpine", "endings", "openingScript"]
    missing = [k for k in required if not parsed.get(k)]
    if missing:
        raise ValueError(f"Story generation response missing required fields: {', '.join(missing)}")

    if len(parsed["characters"]) < 3:
        logger.warning("Only %d characters generated, expected 3-5", len(parsed["characters"]))
    if len(parsed["storySpine"]) != 10:
        logger.warning("storySpine has %d beats, expected 10", len(parsed["storySpine"]))
    if len(parsed["endings"]) != 5:
        logger.warning("endings has %d entries, expected 5", len(parsed["endings"]))

    # Validate every beat's sceneId + castIds, beat choices, every ending's
    # finalSceneId. Surfacing this here means the pipeline fails fast instead
    # of at runtime when a player walks into a beat that points at a missing
    # asset or has no choices to offer.
    scene_ids = {s["id"] for s in parsed["scenes"]}
    char_ids = {c["id"] for c in parsed["characters"]}
    ending_tags = {e.get("alignmentTag") for e in parsed["endings"] if e.get("alignmentTag")}
    for beat in parsed["storySpine"]:
        if beat.get("sceneId") not in scene_ids:
            logger.warning("beat %s references missing sceneId %s", beat.get("id"), beat.get("sceneId"))
        for cid in beat.get("castIds") or []:
            if cid not in char_ids:
                logger.warning("beat %s references unknown castId %s", beat.get("id"), cid)
        choices = beat.get("choices") or []
        if len(choices) < 2:
            logger.warning("beat %s has only %d choices, expected 3", beat.get("id"), len(choices))
        for ch in choices:
            if ch.get("alignmentTag") not in ending_tags:
                logger.warning("beat %s choice %r references unknown alignmentTag %s",
                               beat.get("id"), (ch.get("text") or "")[:40], ch.get("alignmentTag"))
    for ending in parsed["endings"]:
        if ending.get("finalSceneId") not in scene_ids:
            logger.warning("ending %s references missing finalSceneId %s",
                           ending.get("id"), ending.get("finalSceneId"))

    logger.info(
        "World generated: %s, %d chars, %d scenes, %d beats, %d endings",
        parsed["worldLore"].get("name"),
        len(parsed["characters"]),
        len(parsed["scenes"]),
        len(parsed["storySpine"]),
        len(parsed["endings"]),
    )
    return parsed
